rust/worker/src/distance/types.rs

```diff
@@ -0,0 +1,145 @@
+use crate::errors::{ChromaError, ErrorCodes};
+use thiserror::Error;
+
+/// The distance function enum.
+/// # Description
+/// This enum defines the distance functions supported by indices in Chroma.
+/// # Variants
+/// - `Euclidean` - The Euclidean or l2 norm.
+/// - `Cosine` - The cosine distance. Specifically, 1 - cosine.
+/// - `InnerProduct` - The inner product. Specifically, 1 - inner product.
+/// # Notes
+/// See https://docs.trychroma.com/usage-guide#changing-the-distance-function
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) enum DistanceFunction {
+    Euclidean,
+    Cosine,
+    InnerProduct,
+}
+
+impl DistanceFunction {
+    // TOOD: Should we error if mismatched dimensions?
+    pub(crate) fn distance(&self, a: &[f32], b: &[f32]) -> f32 {
+        // TODO: implement this in SSE/AVX SIMD
+        // For now we write these as loops since we suspect that will more likely
+        // lead to the compiler vectorizing the code. (We saw this on
+        // Apple Silicon Macs who didn't have hand-rolled SIMD instructions in our
+        // C++ code).
+        match self {
+            DistanceFunction::Euclidean => {
+                let mut sum = 0.0;
+                for i in 0..a.len() {
+                    sum += (a[i] - b[i]).powi(2);
+                }
+                sum
+            }
+            DistanceFunction::Cosine => {
+                // For cosine we just assume the vectors have been normalized, since that
+                // is what our indices expect.
+                let mut sum = 0.0;
+                for i in 0..a.len() {
+                    sum += a[i] * b[i];
+                }
+                1.0_f32 - sum
+            }
+            DistanceFunction::InnerProduct => {
+                let mut sum = 0.0;
+                for i in 0..a.len() {
+                    sum += a[i] * b[i];
+                }
+                1.0_f32 - sum
```

Comment:
i'm confused -- if the vectors haven't been normalized then won't this usually be negative? i must be missing something silly

Reply:
I could be missing something...Why would it be negative? You square the sum, so all components are positive - inner product is one when angle is 0. We invert the inner product as per the hnsw libs and python code because we want inner product similarity where 0 == similar and positive = increasingly dissimilar. This matched the intuitive understanding of "distance" and more importantly matches the l2 norm so that all code treats distances the same.